system/mp/test/mptest.cpp

```diff
@@ -521,127 +523,214 @@ void MultiTest(ICommunicator *_comm)
+
+    // ---------
+
+    class Sender : public Thread
     {
-        int m = 0;
-        while (m < (numranks - 1))
+    public:
+        Owned<ICommunicator> mpicomm;
+        int numranks;
+        int myrank;
+        int buffsize;
+        int iters;
+        char *buff;
+        Sender(ICommunicator *_mpicomm, int _numranks, int _myrank, int _buffsize, int _iters)
         {
-            int rankid = 0;
-            while (rankid < numranks)
-            {
-                left = rankid - 1 - m;
-                if (left < 0)
-                    left = numranks + left;
-                right = rankid + 1 + m;
-                if (right >= numranks)
-                    right = right % numranks;
-
-                if (rankid == myrank)
-                {
-                    if (rankid == 0)
-                    {
-                        PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-                        mpicomm->send(mb,right,MPTAG_TEST);
-                        mb2.clear();
-                        PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-                        mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-                        mb2.read(rankval);
-                        ranksum += rankval;
-                    }
-                    else
-                    {
-                        mb2.clear();
-                        PrintLog("MPTEST: MPITest: %d recv from rank %d", myrank, left);
-                        mpicomm->recv(mb2,left,MPTAG_TEST,&r);
-                        mb2.read(rankval);
-                        ranksum += rankval;
-                        PrintLog("MPTEST: MPITest: %d send to rank %d", myrank, right);
-                        mpicomm->send(mb,right,MPTAG_TEST);
-                    }
-                }
-                rankid++;
+            mpicomm.set(_mpicomm);
+            numranks = _numranks;
+            myrank = _myrank;
+            buffsize = _buffsize;
+            iters = _iters;
+        }
+
+        int run()
+        {
+            PrintLog("MPTEST: MPAlltoAll sender started, myrank = %d", myrank);
+
+            int pintvl = iters/10;
+            if (pintvl < 5)
+                pintvl = 5;
 
+            char *buff = new char[buffsize+1];
+            memset(buff, 'a', buffsize);
+            CMessageBuffer smb;
+            smb.append(buffsize, buff);
+
+            for (int k=1;k<=iters;k++)
+            {
+                mpicomm->send(smb,RANK_ALL_OTHER,MPTAG_TEST);
+                if ((k%pintvl) == 0)
+                    PrintLog("MPTEST: MPAlltoAll sender %d iteration %d complete", myrank, k);
             }
-            m++;
+
+            mpicomm->barrier();
+            delete [] buff;
+            PrintLog("MPTEST: MPAlltoAll sender stopped");
+            return 0;
         }
-    }
+    } sender(_mpicomm, numranks, myrank, buffsize, iters);
```

Comment:
I wouldn't change, but you could possibly use a non-blocking send (MP_ASYNC_SEND) before the recv loop each time.